In [38]:
import pandas as pd

# Pre-process MovieLens' Movies dataset

In [39]:
df_ml_movies = pd.read_csv('ml_original/movies.csv')
df_ml_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62423 entries, 0 to 62422
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  62423 non-null  int64 
 1   title    62423 non-null  object
 2   genres   62423 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.4+ MB


In [40]:
df_ml_movies.head(10)

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   
5        6                         Heat (1995)   
6        7                      Sabrina (1995)   
7        8                 Tom and Huck (1995)   
8        9                 Sudden Death (1995)   
9       10                    GoldenEye (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy  
5                        Action|Crime|Thriller  
6                               Comedy|Romance  
7                           Adventure|Children  
8                                       Action  
9                    Action|Adventure|Thriller

In [41]:
df_ml_movies[df_ml_movies['title'] == 'Inception (2010)']

movieId             title  \
14937    79132  Inception (2010)   

                                                genres  
14937  Action|Crime|Drama|Mystery|Sci-Fi|Thriller|IMAX

In [42]:
df_ml_movies = df_ml_movies[['movieId', 'title']]
df_ml_movies.head()
# drop genre

movieId                               title
0        1                    Toy Story (1995)
1        2                      Jumanji (1995)
2        3             Grumpier Old Men (1995)
3        4            Waiting to Exhale (1995)
4        5  Father of the Bride Part II (1995)

# Standardise movie titles that start with "The"

In [43]:
df_ml_movies[df_ml_movies['title'].str.contains('Shawshank')]
#need to convert all movies with 'The'

movieId                             title
314      318  Shawshank Redemption, The (1994)

In [44]:
print(df_ml_movies[df_ml_movies['title'].str.contains(', The \(\d+')].head(50))
df_ml_with_the = df_ml_movies[df_ml_movies['title'].str.contains(', The \(\d+')]
df_ml_with_the

     movieId                                              title
10        11                     American President, The (1995)
49        50                         Usual Suspects, The (1995)
53        54                              Big Green, The (1995)
59        60                 Indian in the Cupboard, The (1995)
77        78                         Crossing Guard, The (1995)
78        79                                  Juror, The (1996)
103      105              Bridges of Madison County, The (1995)
115      117              Young Poisoner's Handbook, The (1995)
119      121                    Boys of St. Vincent, The (1992)
124      126                  NeverEnding Story III, The (1994)
136      138                             Neon Bible, The (1995)
139      141                               Birdcage, The (1996)
142      144                      Brothers McMullen, The (1995)
144      146                Amazing Panda Adventure, The (1995)
145      147                     Basketb

movieId                                          title
10          11                 American President, The (1995)
49          50                     Usual Suspects, The (1995)
53          54                          Big Green, The (1995)
59          60             Indian in the Cupboard, The (1995)
77          78                     Crossing Guard, The (1995)
...        ...                                            ...
34765   146202  Inside of the White Slave Traffic, The (1913)
35759   148626                          Big Short, The (2015)
38533   155026                        Office Wife, The (1930)
39625   157625                  Whispering Chorus, The (1918)
40340   159185               Extraordinary Voyage, The (2011)

[3405 rows x 2 columns]

In [45]:
df_ml_without_the = df_ml_movies[~df_ml_movies['title'].str.contains(', The \(\d+')]
df_ml_without_the

movieId                               title
0            1                    Toy Story (1995)
1            2                      Jumanji (1995)
2            3             Grumpier Old Men (1995)
3            4            Waiting to Exhale (1995)
4            5  Father of the Bride Part II (1995)
...        ...                                 ...
62418   209157                           We (2018)
62419   209159           Window of the Soul (2001)
62420   209163                    Bad Poems (2018)
62421   209169                 A Girl Thing (2001)
62422   209171      Women of Devil's Island (1962)

[59018 rows x 2 columns]

In [46]:
df_test = pd.concat([df_ml_without_the, df_ml_with_the])
df_test

movieId                                          title
0            1                               Toy Story (1995)
1            2                                 Jumanji (1995)
2            3                        Grumpier Old Men (1995)
3            4                       Waiting to Exhale (1995)
4            5             Father of the Bride Part II (1995)
...        ...                                            ...
34765   146202  Inside of the White Slave Traffic, The (1913)
35759   148626                          Big Short, The (2015)
38533   155026                        Office Wife, The (1930)
39625   157625                  Whispering Chorus, The (1918)
40340   159185               Extraordinary Voyage, The (2011)

[62423 rows x 2 columns]

In [47]:
raw_the = df_ml_with_the['title'].str.split(', The ',)
print(raw_the.head(15))
df_ml_with_the['new'] = 'The ' + raw_the.str[0] + ' ' + raw_the.str[1]
df_ml_with_the.head()

10            [American President, (1995)]
49                [Usual Suspects, (1995)]
53                     [Big Green, (1995)]
59        [Indian in the Cupboard, (1995)]
77                [Crossing Guard, (1995)]
78                         [Juror, (1996)]
103    [Bridges of Madison County, (1995)]
115    [Young Poisoner's Handbook, (1995)]
119          [Boys of St. Vincent, (1992)]
124        [NeverEnding Story III, (1994)]
136                   [Neon Bible, (1995)]
139                     [Birdcage, (1996)]
142            [Brothers McMullen, (1995)]
144      [Amazing Panda Adventure, (1995)]
145           [Basketball Diaries, (1995)]
Name: title, dtype: object


C:\Users\Kevan\AppData\Local\Temp/ipykernel_16824/2976948629.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ml_with_the['new'] = 'The ' + raw_the.str[0] + ' ' + raw_the.str[1]


movieId                               title  \
10       11      American President, The (1995)   
49       50          Usual Suspects, The (1995)   
53       54               Big Green, The (1995)   
59       60  Indian in the Cupboard, The (1995)   
77       78          Crossing Guard, The (1995)   

                                  new  
10      The American President (1995)  
49          The Usual Suspects (1995)  
53               The Big Green (1995)  
59  The Indian in the Cupboard (1995)  
77          The Crossing Guard (1995)

In [48]:
df_ml_with_the.drop('new', axis=1, inplace=True)

C:\Users\Kevan\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [49]:
df_ml_with_the['title'] = 'The ' + raw_the.str[0] + ' ' + raw_the.str[1]
# df_ml_with_the = df_ml_with_the.drop('new', axis=1)
df_ml_with_the.head()

C:\Users\Kevan\AppData\Local\Temp/ipykernel_16824/889700230.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ml_with_the['title'] = 'The ' + raw_the.str[0] + ' ' + raw_the.str[1]


movieId                              title
10       11      The American President (1995)
49       50          The Usual Suspects (1995)
53       54               The Big Green (1995)
59       60  The Indian in the Cupboard (1995)
77       78          The Crossing Guard (1995)

In [75]:
df_ml_movies = pd.concat([df_ml_without_the, df_ml_with_the])
df_ml_movies.sort_values(by=['movieId'], inplace=True)
df_ml_movies.head(15)
# We can see The American President (1995) on row 10 being in the correct format now

movieId                               title
0         1                    Toy Story (1995)
1         2                      Jumanji (1995)
2         3             Grumpier Old Men (1995)
3         4            Waiting to Exhale (1995)
4         5  Father of the Bride Part II (1995)
5         6                         Heat (1995)
6         7                      Sabrina (1995)
7         8                 Tom and Huck (1995)
8         9                 Sudden Death (1995)
9        10                    GoldenEye (1995)
10       11       The American President (1995)
11       12  Dracula: Dead and Loving It (1995)
12       13                        Balto (1995)
13       14                        Nixon (1995)
14       15             Cutthroat Island (1995)

# Pre-Processing IMDB Web Scraped dataset

In [51]:
df_scraped = pd.read_csv('scraped_cleaned/scraped_one_hot_encoded.csv', dtype={'num_of_rating': str}, index_col=0)
df_scraped.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 113642 entries, 0 to 113642
Data columns (total 95 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Unnamed: 0.1     113642 non-null  int64  
 1   title            113642 non-null  object 
 2   runtime          106500 non-null  object 
 3   rating           113641 non-null  float64
 4   num_of_rating    113641 non-null  object 
 5   budget           21383 non-null   object 
 6   worldwide_gross  27521 non-null   object 
 7   origin           113182 non-null  object 
 8   cast             113642 non-null  object 
 9   director         113642 non-null  object 
 10  writer           113642 non-null  object 
 11  producer         113642 non-null  object 
 12  composer         113642 non-null  object 
 13  cinematographer  113642 non-null  object 
 14  editor           113642 non-null  object 
 15  year             84242 non-null   float64
 16  Action           113642 non-null  int6

In [52]:
df_scraped.head(10)

Unnamed: 0.1                                              title  runtime  \
0             0                    The Shawshank Redemption (1994)  142 min   
1             1                             The Dark Knight (2008)  152 min   
2             2                                   Inception (2010)  148 min   
3             3                                  Fight Club (1999)  139 min   
4             4                                Pulp Fiction (1994)  154 min   
5             5                                Forrest Gump (1994)  142 min   
6             6                                  The Matrix (1999)  136 min   
7             7                           Breaking Bad (2008–2013)   49 min   
8             8  The Lord of the Rings: The Fellowship of the R...  178 min   
9             9                        Game of Thrones (2011–2019)   57 min   

   rating num_of_rating                    budget worldwide_gross  \
0     9.3          2.7M   $25,000,000 (estimated)     $28,884,504   
1     9.0          2.7M  $185,000,000 (estimated)  $1,006,234,167   
2     8.8          2.4M  $160,000,000 (estimated)    $836,848,102   
3     8.8          2.1M   $63,000,000 (estimated)    $101,209,702   
4     8.9          2.1M    $8,000,000 (estimated)    $213,928,762   
5     8.8          2.1M   $55,000,000 (estimated)    $678,226,465   
6     8.7          1.9M   $63,000,000 (estimated)    $467,222,728   
7     9.5          1.9M                       NaN             NaN   
8     8.8          1.9M   $93,000,000 (estimated)    $898,204,420   
9     9.2          2.1M                       NaN             NaN   

                         origin  \
0                 United States   
1  United States,United Kingdom   
2  United States,United Kingdom   
3         Germany,United States   
4                 United States   
5                 United States   
6       United States,Australia   
7                 United States   
8     New Zealand,United States   
9  United States,United Kingdom   

                                                cast  \
0  ['Tim Robbins,Morgan Freeman,Bob Gunton,Willia...   
1  ["Christian Bale,Heath Ledger,Aaron Eckhart,Mi...   
2  ['Leonardo DiCaprio,Joseph Gordon-Levitt,Ellio...   
3  ["Edward Norton,Brad Pitt,Meat Loaf,Zach Greni...   
4  ['Tim Roth,Amanda Plummer,Laura Lovelace,John ...   
5  ["Tom Hanks,Rebecca Williams,Sally Field,Micha...   
6  ["Keanu Reeves,Laurence Fishburne,Carrie-Anne ...   
7  ["Bryan Cranston,Anna Gunn,Aaron Paul,Betsy Br...   
8  ["Alan Howard,Noel Appleby,Sean Astin,Sala Bak...   
9  ["Peter Dinklage,Lena Headey,Emilia Clarke,Kit...   

                                            director  ... TV-G TV-MA TV-PG  \
0                                 ['Frank Darabont']  ...    0     0     0   
1                              ['Christopher Nolan']  ...    0     0     0   
2                              ['Christopher Nolan']  ...    0     0     0   
3                                  ['David Fincher']  ...    0     0     0   
4                              ['Quentin Tarantino']  ...    0     0     0   
5                                ['Robert Zemeckis']  ...    0     0     0   
6                 ['Lana Wachowski,Lilly Wachowski']  ...    0     0     0   
7  ['Michelle MacLaren,Adam Bernstein,Vince Gilli...  ...    0     1     0   
8                                  ['Peter Jackson']  ...    0     0     0   
9  ['David Nutter,Alan Taylor,Alex Graves,Miguel ...  ...    0     1     0   

  TV-Y TV-Y7  TV-Y7-FV  U/A  Unrated  X  certificate_nan  
0    0     0         0    0        0  0                0  
1    0     0         0    0        0  0                0  
2    0     0         0    0        0  0                0  
3    0     0         0    0        0  0                0  
4    0     0         0    0        0  0                0  
5    0     0         0    0        0  0                0  
6    0     0         0    0        0  0                0  
7    0     0         0    0        0  0

In [53]:
# find rows which are NOT TV shows
df_scraped[~df_scraped.title.str.contains('–')].head(15)

Unnamed: 0.1                                              title  runtime  \
0              0                    The Shawshank Redemption (1994)  142 min   
1              1                             The Dark Knight (2008)  152 min   
2              2                                   Inception (2010)  148 min   
3              3                                  Fight Club (1999)  139 min   
4              4                                Pulp Fiction (1994)  154 min   
5              5                                Forrest Gump (1994)  142 min   
6              6                                  The Matrix (1999)  136 min   
8              8  The Lord of the Rings: The Fellowship of the R...  178 min   
10            10                               The Godfather (1972)  175 min   
11            11  The Lord of the Rings: The Return of the King ...  201 min   
12            12                                Interstellar (2014)  169 min   
13            13       The Lord of the Rings: The Two Towers (2002)  179 min   
14            14                       The Dark Knight Rises (2012)  164 min   
15            15                                       Se7en (1995)  127 min   
16            16                            Django Unchained (2012)  165 min   

    rating num_of_rating                    budget worldwide_gross  \
0      9.3          2.7M   $25,000,000 (estimated)     $28,884,504   
1      9.0          2.7M  $185,000,000 (estimated)  $1,006,234,167   
2      8.8          2.4M  $160,000,000 (estimated)    $836,848,102   
3      8.8          2.1M   $63,000,000 (estimated)    $101,209,702   
4      8.9          2.1M    $8,000,000 (estimated)    $213,928,762   
5      8.8          2.1M   $55,000,000 (estimated)    $678,226,465   
6      8.7          1.9M   $63,000,000 (estimated)    $467,222,728   
8      8.8          1.9M   $93,000,000 (estimated)    $898,204,420   
10     9.2          1.9M    $6,000,000 (estimated)    $250,341,816   
11     9.0          1.9M   $94,000,000 (estimated)  $1,146,457,748   
12     8.6          1.8M  $165,000,000 (estimated)    $773,867,216   
13     8.8          1.7M   $94,000,000 (estimated)    $947,944,270   
14     8.4          1.7M  $250,000,000 (estimated)  $1,081,169,825   
15     8.6          1.7M   $33,000,000 (estimated)    $327,333,559   
16     8.4          1.6M  $100,000,000 (estimated)    $426,074,373   

                                 origin  \
0                         United States   
1          United States,United Kingdom   
2          United States,United Kingdom   
3                 Germany,United States   
4                         United States   
5                         United States   
6               United States,Australia   
8             New Zealand,United States   
10                        United States   
11            New Zealand,United States   
12  United States,United Kingdom,Canada   
13            New Zealand,United States   
14         United States,United Kingdom   
15                        United States   
16                        United States   

                                                 cast  \
0   ['Tim Robbins,Morgan Freeman,Bob Gunton,Willia...   
1   ["Christian Bale,Heath Ledger,Aaron Eckhart,Mi...   
2   ['Leonardo DiCaprio,Joseph Gordon-Levitt,Ellio...   
3   ["Edward Norton,Brad Pitt,Meat Loaf,Zach Greni...   
4   ['Tim Roth,Amanda Plummer,Laura Lovelace,John ...   
5   ["Tom Hanks,Rebecca Williams,Sally Field,Micha...   
6   ["Keanu Reeves,Laurence Fishburne,Carrie-Anne ...   
8   ["Alan Howard,Noel Appleby,Sean Astin,Sala Bak...   
10  ['Marlon Brando,Al Pacino,James Caan,Richard S...   
11  ['Noel Appleby,Ali Astin,Sean Astin,David Asto...   
12  ['Ellen Burstyn,Matthew McConaughey,Mackenzie ...   
13  ['Bruce Allpress,Sean Astin,John Bach,Sala Bak...   
14  ["Christian Bale,Gary Oldman,Tom Hardy,Joseph ...   
15  ['Morgan Freeman,Andrew Kevin Walker,Daniel Za...   
16  ["Jamie Foxx,Christoph Waltz,Leonardo DiCaprio...   



In [54]:
# remove all TV shows
df_scraped = df_scraped[~df_scraped.title.str.contains('–')]
df_scraped.info()
# we can see that the movies we have dropped from 113k to 100k

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100791 entries, 0 to 113640
Data columns (total 95 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Unnamed: 0.1     100791 non-null  int64  
 1   title            100791 non-null  object 
 2   runtime          96534 non-null   object 
 3   rating           100791 non-null  float64
 4   num_of_rating    100791 non-null  object 
 5   budget           21383 non-null   object 
 6   worldwide_gross  27521 non-null   object 
 7   origin           100476 non-null  object 
 8   cast             100791 non-null  object 
 9   director         100791 non-null  object 
 10  writer           100791 non-null  object 
 11  producer         100791 non-null  object 
 12  composer         100791 non-null  object 
 13  cinematographer  100791 non-null  object 
 14  editor           100791 non-null  object 
 15  year             84242 non-null   float64
 16  Action           100791 non-null  int6

In [55]:
df_scraped[df_scraped['title'] == 'Inception (2010)']

Unnamed: 0.1             title  runtime  rating num_of_rating  \
2             2  Inception (2010)  148 min     8.8          2.4M   

                     budget worldwide_gross                        origin  \
2  $160,000,000 (estimated)    $836,848,102  United States,United Kingdom   

                                                cast               director  \
2  ['Leonardo DiCaprio,Joseph Gordon-Levitt,Ellio...  ['Christopher Nolan']   

   ... TV-G TV-MA TV-PG TV-Y TV-Y7  TV-Y7-FV  U/A  Unrated  X  certificate_nan  
2  ...    0     0     0    0     0         0    0        0  0                0  

[1 rows x 95 columns]

In [56]:
df_ml_movies[df_ml_movies['title'].str.contains('Shawshank')]

movieId                            title
314      318  The Shawshank Redemption (1994)

In [57]:
df_scraped[df_scraped['title'].str.contains('Shawshank')]

Unnamed: 0.1                                              title  \
0                  0                    The Shawshank Redemption (1994)   
105064        105064  The Shawshank Redemption: Behind the Scenes (2...   

        runtime  rating num_of_rating                   budget  \
0       142 min     9.3          2.7M  $25,000,000 (estimated)   
105064    9 min     9.3           208                      NaN   

       worldwide_gross          origin  \
0          $28,884,504   United States   
105064             NaN  United Kingdom   

                                                     cast            director  \
0       ['Tim Robbins,Morgan Freeman,Bob Gunton,Willia...  ['Frank Darabont']   
105064      ['Frank Darabont,Morgan Freeman,Tim Robbins']             ['nan']   

        ... TV-G TV-MA TV-PG TV-Y TV-Y7  TV-Y7-FV  U/A  Unrated  X  \
0       ...    0     0     0    0     0         0    0        0  0   
105064  ...    0     0     0    0     0         0    0        0  0   

        certificate_nan  
0                     0  
105064                1  

[2 rows x 95 columns]

In [58]:
titles_merged = pd.merge(df_scraped, df_ml_movies)
titles_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22744 entries, 0 to 22743
Data columns (total 96 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0.1     22744 non-null  int64  
 1   title            22744 non-null  object 
 2   runtime          22722 non-null  object 
 3   rating           22744 non-null  float64
 4   num_of_rating    22744 non-null  object 
 5   budget           8347 non-null   object 
 6   worldwide_gross  9264 non-null   object 
 7   origin           22731 non-null  object 
 8   cast             22744 non-null  object 
 9   director         22744 non-null  object 
 10  writer           22744 non-null  object 
 11  producer         22744 non-null  object 
 12  composer         22744 non-null  object 
 13  cinematographer  22744 non-null  object 
 14  editor           22744 non-null  object 
 15  year             22744 non-null  float64
 16  Action           22744 non-null  int64  
 17  Adult       

## The number of movies that exist in BOTH MovieLens and IMDB Web Scraped is 22,744. Only 8,347 of them have the budget data

In [59]:
titles_merged.head(30)

Unnamed: 0.1                                              title  runtime  \
0              0                    The Shawshank Redemption (1994)  142 min   
1              1                             The Dark Knight (2008)  152 min   
2              2                                   Inception (2010)  148 min   
3              3                                  Fight Club (1999)  139 min   
4              4                                Pulp Fiction (1994)  154 min   
5              5                                Forrest Gump (1994)  142 min   
6              6                                  The Matrix (1999)  136 min   
7              8  The Lord of the Rings: The Fellowship of the R...  178 min   
8             10                               The Godfather (1972)  175 min   
9             11  The Lord of the Rings: The Return of the King ...  201 min   
10            12                                Interstellar (2014)  169 min   
11            13       The Lord of the Rings: The Two Towers (2002)  179 min   
12            14                       The Dark Knight Rises (2012)  164 min   
13            16                            Django Unchained (2012)  165 min   
14            17                                   Gladiator (2000)  155 min   
15            18                               Batman Begins (2005)  140 min   
16            19                        Inglourious Basterds (2009)  153 min   
17            20                    The Silence of the Lambs (1991)  118 min   
18            21                     The Wolf of Wall Street (2013)  180 min   
19            22                         Saving Private Ryan (1998)  169 min   
20            24                            Schindler's List (1993)  195 min   
21            25                                The Avengers (2012)  143 min   
22            26                                The Prestige (2006)  130 min   
23            27                                The Departed (2006)  151 min   
24            28                              Shutter Island (2010)  138 min   
25            29                              The Green Mile (1999)  189 min   
26            31                                      Avatar (2009)  162 min   
27            33                                     Memento (2000)  113 min   
28            35                          Back to the Future (1985)  116 min   
29            36                     Guardians of the Galaxy (2014)  121 min   

    rating num_of_rating                    budget worldwide_gross  \
0      9.3          2.7M   $25,000,000 (estimated)     $28,884,504   
1      9.0          2.7M  $185,000,000 (estimated)  $1,006,234,167   
2      8.8          2.4M  $160,000,000 (estimated)    $836,848,102   
3      8.8          2.1M   $63,000,000 (estimated)    $101,209,702   
4      8.9          2.1M    $8,000,000 (estimated)    $213,928,762   
5      8.8          2.1M   $55,000,000 (estimated)    $678,226,465   
6      8.7          1.9M   $63,000,000 (estimated)    $467,222,728   
7      8.8          1.9M   $93,000,000 (estimated)    $898,204,420   
8      9.2          1.9M    $6,000,000 (estimated)    $250,341,816   
9      9.0          1.9M   $94,000,000 (estimated)  $1,146,457,748   
10     8.6          1.8M  $165,000,000 (estimated)    $773,867,216   
11     8.8          1.7M   $94,000,000 (estimated)    $947,944,270   
12     8.4          1.7M  $250,000,000 (estimated)  $1,081,169,825   
13     8.4          1.6M  $100,000,000 (estimated)    $426,074,373   
14     8.5          1.5M  $103,000,000 (estimated)    $503,162,313   
15     8.2          1.5M  $150,000,000 (estimated)    $373,672,993   
16     8.3          1.5M   $70,000,000 (estimated)    $321,457,747   
17     8.6          1.4M   $19,000,000 (estimated)    $272,742,922   
18     8.2          1.4M  $100,000,000 (estimated)    $406,878,233   
19     8.6          1.4M   $70,000,000 (estimated)    $482,349,603   
20     9.0          1.4M   $22,000,000 (estimated)   

In [60]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None): 
    display(titles_merged.head())

Unnamed: 0.1                            title  runtime  rating  \
0             0  The Shawshank Redemption (1994)  142 min     9.3   
1             1           The Dark Knight (2008)  152 min     9.0   
2             2                 Inception (2010)  148 min     8.8   
3             3                Fight Club (1999)  139 min     8.8   
4             4              Pulp Fiction (1994)  154 min     8.9   

  num_of_rating                    budget worldwide_gross  \
0          2.7M   $25,000,000 (estimated)     $28,884,504   
1          2.7M  $185,000,000 (estimated)  $1,006,234,167   
2          2.4M  $160,000,000 (estimated)    $836,848,102   
3          2.1M   $63,000,000 (estimated)    $101,209,702   
4          2.1M    $8,000,000 (estimated)    $213,928,762   

                         origin  \
0                 United States   
1  United States,United Kingdom   
2  United States,United Kingdom   
3         Germany,United States   
4                 United States   

                                                cast               director  \
0  ['Tim Robbins,Morgan Freeman,Bob Gunton,Willia...     ['Frank Darabont']   
1  ["Christian Bale,Heath Ledger,Aaron Eckhart,Mi...  ['Christopher Nolan']   
2  ['Leonardo DiCaprio,Joseph Gordon-Levitt,Ellio...  ['Christopher Nolan']   
3  ["Edward Norton,Brad Pitt,Meat Loaf,Zach Greni...      ['David Fincher']   
4  ['Tim Roth,Amanda Plummer,Laura Lovelace,John ...  ['Quentin Tarantino']   

                                              writer  \
0                    ['Stephen King,Frank Darabont']   
1  ['Jonathan Nolan,Christopher Nolan,Christopher...   
2                              ['Christopher Nolan']   
3                       ['Chuck Palahniuk,Jim Uhls']   
4  ['Quentin Tarantino,Roger Avary,Quentin Tarant...   

                                            producer  \
0        ['Liz Glotzer,David V. Lester,Niki Marvin']   
1  ['Kevin de la Noy,Jordan Goldberg,Philip Lee,B...   
2  ['Zakaria Alaoui,John Bernard,Chris Brigham,Jo...   
3  ['Ross Grayson Bell,Ceán Chaffin,John S. Dorse...   
4  ['Lawrence Bender,Danny DeVito,Richard N. Glad...   

                                      composer      cinematographer  \
0                            ['Thomas Newman']    ['Roger Deakins']   
1          ['James Newton Howard,Hans Zimmer']    ['Wally Pfister']   
2                              ['Hans Zimmer']    ['Wally Pfister']   
3  ['Dust Brothers,John King,Michael Simpson']  ['Jeff Cronenweth']   
4                                      ['nan']   ['Andrzej Sekula']   

                      editor    year  Action  Adult  Adventure  Animation  \
0  ['Richard Francis-Bruce']  1990.0       0      0          0          0   
1              ['Lee Smith']  2000.0       1      0          0          0   
2              ['Lee Smith']  2000.0       1      0          1          0   
3          ['James Haygood']  1990.0       0      0          0          0   
4            ['Sally Menke']  1990.0       0      0          0          0   

   Biography  Comedy  Crime  Documentary  Drama  Family  Fantasy  Film-Noir  \
0          0       0      0            0      1       0        0          0   
1          0       0      1            0      1       0        0          0   
2          0       0      0            0      0       0        0          0   
3          0       0      0            0      1       0        0          0   
4          0       0      1            0      1       0        0          0   

   Game-Show  History  Horror  Music  Musical  Mystery  News  Reality-TV  \
0          0        0       0      0        0        0     0           0   
1          0        0       0      0        0        0     0           0   
2          0        0       0      0        0        0     0           0   
3          0        0       0      0        0        0     0           0   
4          0        0       0      0        0        0     0           0   

   Romance  Sci-Fi  Short  Sport  Talk-

In [61]:
titles_merged[titles_merged['title'] == 'Inception (2010)']

Unnamed: 0.1             title  runtime  rating num_of_rating  \
2             2  Inception (2010)  148 min     8.8          2.4M   

                     budget worldwide_gross                        origin  \
2  $160,000,000 (estimated)    $836,848,102  United States,United Kingdom   

                                                cast               director  \
2  ['Leonardo DiCaprio,Joseph Gordon-Levitt,Ellio...  ['Christopher Nolan']   

   ... TV-MA TV-PG TV-Y TV-Y7 TV-Y7-FV  U/A  Unrated  X  certificate_nan  \
2  ...     0     0    0     0        0    0        0  0                0   

   movieId  
2    79132  

[1 rows x 96 columns]

In [62]:
titles_merged_movieId_only = titles_merged[['movieId']]
titles_merged_movieId_only.head()

movieId
0      318
1    58559
2    79132
3     2959
4      296

# Pre-processing MovieLens' ratings dataset

In [77]:
df_ml_ratings = pd.read_csv('ml_original/ratings.csv')
df_ml_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000095 entries, 0 to 25000094
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 762.9 MB


In [78]:
df_ml_ratings.drop('timestamp', axis=1, inplace=True)
df_ml_ratings

userId  movieId  rating
0              1      296     5.0
1              1      306     3.5
2              1      307     5.0
3              1      665     5.0
4              1      899     3.5
...          ...      ...     ...
25000090  162541    50872     4.5
25000091  162541    55768     2.5
25000092  162541    56176     2.0
25000093  162541    58559     4.0
25000094  162541    63876     5.0

[25000095 rows x 3 columns]

In [79]:
len(df_ml_ratings['movieId'].tolist())
# originally, we have around 25M ratings dataset

25000095

# Finding out the final number of data we have

In [86]:
final_number = df_ml_ratings[df_ml_ratings['movieId'].isin(titles_merged['movieId'].tolist())]

In [87]:
final_number.info()
print()
print(final_number.head())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13532791 entries, 0 to 25000093
Data columns (total 3 columns):
 #   Column   Dtype  
---  ------   -----  
 0   userId   int64  
 1   movieId  int64  
 2   rating   float64
dtypes: float64(1), int64(2)
memory usage: 413.0 MB

    userId  movieId  rating
0        1      296     5.0
3        1      665     5.0
7        1     1217     3.5
9        1     1250     4.0
12       1     2011     2.5


In [88]:
final_number.nunique()

userId     162541
movieId     21567
rating         10
dtype: int64

# We have:
# 13,532,791 ratings
# 162,541 unique userId
# 21,567 unique movies

# FINAL DATAFRAMES WE GONNA USE
# df_movies and df_ratings

In [89]:
df_movies = titles_merged
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22744 entries, 0 to 22743
Data columns (total 96 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0.1     22744 non-null  int64  
 1   title            22744 non-null  object 
 2   runtime          22722 non-null  object 
 3   rating           22744 non-null  float64
 4   num_of_rating    22744 non-null  object 
 5   budget           8347 non-null   object 
 6   worldwide_gross  9264 non-null   object 
 7   origin           22731 non-null  object 
 8   cast             22744 non-null  object 
 9   director         22744 non-null  object 
 10  writer           22744 non-null  object 
 11  producer         22744 non-null  object 
 12  composer         22744 non-null  object 
 13  cinematographer  22744 non-null  object 
 14  editor           22744 non-null  object 
 15  year             22744 non-null  float64
 16  Action           22744 non-null  int64  
 17  Adult       

In [90]:
df_movies

Unnamed: 0.1                                   title  runtime  rating  \
0                 0         The Shawshank Redemption (1994)  142 min     9.3   
1                 1                  The Dark Knight (2008)  152 min     9.0   
2                 2                        Inception (2010)  148 min     8.8   
3                 3                       Fight Club (1999)  139 min     8.8   
4                 4                     Pulp Fiction (1994)  154 min     8.9   
...             ...                                     ...      ...     ...   
22739        113506  Doggiewoggiez! Poochiewoochiez! (2012)   55 min     6.8   
22740        113558                        Bobbikins (1959)   89 min     5.0   
22741        113574                           Vallen (2001)  105 min     5.8   
22742        113590                  Exclusive Story (1936)   73 min     6.0   
22743        113618               His Wooden Wedding (1925)   20 min     7.4   

      num_of_rating                    budget worldwide_gross  \
0              2.7M   $25,000,000 (estimated)     $28,884,504   
1              2.7M  $185,000,000 (estimated)  $1,006,234,167   
2              2.4M  $160,000,000 (estimated)    $836,848,102   
3              2.1M   $63,000,000 (estimated)    $101,209,702   
4              2.1M    $8,000,000 (estimated)    $213,928,762   
...             ...                       ...             ...   
22739           182        $2,000 (estimated)             NaN   
22740           182                       NaN             NaN   
22741           182                       NaN             NaN   
22742           182                       NaN             NaN   
22743           182                       NaN             NaN   

                             origin  \
0                     United States   
1      United States,United Kingdom   
2      United States,United Kingdom   
3             Germany,United States   
4                     United States   
...                             ...   
22739                 United States   
22740                United Kingdom   
22741                       Belgium   
22742                 United States   
22743                 United States   

                                                    cast  \
0      ['Tim Robbins,Morgan Freeman,Bob Gunton,Willia...   
1      ["Christian Bale,Heath Ledger,Aaron Eckhart,Mi...   
2      ['Leonardo DiCaprio,Joseph Gordon-Levitt,Ellio...   
3      ["Edward Norton,Brad Pitt,Meat Loaf,Zach Greni...   
4      ['Tim Roth,Amanda Plummer,Laura Lovelace,John ...   
...                                                  ...   
22739  ['Tim Allen,Michael Badalucco,Jim Belushi,Matt...   
22740  ["Shirley Jones,Max Bygraves,Billie Whitelaw,B...   
22741  ['Lee Williams,Emma Thomas,Alice Krige,Lydia C...   
22742  ["Franchot Tone,Madge Evans,Stuart Erwin,Josep...   
22743  ["Charley Chase,Katherine Grant,Gale Henry,Fre...   

                    director  ... TV-MA TV-PG TV-Y TV-Y7 TV-Y7-FV  U/A  \
0         ['Frank Darabont']  ...     0     0    0     0        0    0   
1      ['Christopher Nolan']  ...     0     0    0     0        0    0   
2      ['Christopher Nolan']  ...     0     0    0     0        0    0   
3          ['David Fincher']  ...     0     0    0     0        0    0   
4      ['Quentin Tarantino']  ...     0     0    0     0        0    0   
...                      ...  ...   ...   ...  ...   ...      ...  ...   
22739                ['nan']  ...     0     0    0     0        0    0   
22740         ['Robert Day']  ...     0     0    0     0        0    0   
22741       ['Hans Herbots']  ...     0     0    0     0        0    0   
22742    ['George B. Seitz']  ...     0     0    0     0        0    0   
22743        ['Leo McCarey']  ...     0     0    0     0        0    0   

       Unrated  X  certificate_nan  movieId  
0            0  0                0      318  
1            0  0                0    58559  
2            0  0                0    79132  
3 

In [93]:
df_ratings = final_number

In [94]:
df_ratings.sort_values(by='movieId', inplace=True)

C:\Users\Kevan\anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [96]:
df_ratings

userId  movieId  rating
19389804  125865        1     3.0
1088714     7345        1     4.5
2060498    13725        1     4.0
11049761   71846        1     3.0
9887084    64189        1     4.5
...          ...      ...     ...
6048882    39183   208941     3.0
14274070   92412   208943     2.0
13009941   84238   209041     3.0
2278203    15152   209053     3.5
22421588  145795   209145     1.0

[13532791 rows x 3 columns]